subsys/bluetooth/host/gatt.c

```diff
@@ -1211,73 +1228,135 @@ static void clear_ccc_cfg(struct bt_gatt_ccc_cfg *cfg)
+static void gatt_delayed_store_free(struct ds_peer *el)
 {
-	if ((!gatt_ccc_conn_is_queued(conn)) &&
-	    bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
-		/* Store the connection with the same index it has in
-		 * the conns array
-		 */
-		gatt_ccc_store.conn_list[bt_conn_index(conn)] =
-			bt_conn_ref(conn);
+	if (el) {
+		memset(&el->peer, 0, sizeof(el->peer));
+		atomic_set(el->flags, 0);
+		LOG_ERR("free");
+	}
+}
+
+static void gatt_delayed_store_enqueue(uint8_t id, const bt_addr_le_t *peer_addr,
+				       enum delayed_store_flags flag)
+{
+	bool bonded = bt_addr_le_is_bonded(id, peer_addr);
+	struct ds_peer *el = gatt_delayed_store_find(id, peer_addr);
+
+	if (bonded) {
+		if (el == NULL) {
+			el = gatt_delayed_store_alloc(id, peer_addr);
+			if (el == NULL) {
+				LOG_ERR("Can't save CF / CCC to flash");
+				return;
+			}
+		}
 
-		k_work_reschedule(&gatt_ccc_store.work, CCC_STORE_DELAY);
+		atomic_set_bit(el->flags, flag);
+
+		LOG_ERR("enqueue");
+		k_work_reschedule(&gatt_delayed_store.work,
+				  K_MSEC(CONFIG_BT_SETTINGS_DELAYED_STORE_MS));
 	}
 }
 
-static bool gatt_ccc_conn_queue_is_empty(void)
+static bool gatt_delayed_work_queue_is_empty(void)
 {
-	for (size_t i = 0; i < CONFIG_BT_MAX_CONN; i++) {
-		if (gatt_ccc_store.conn_list[i]) {
+	for (int i = 0; i < ARRAY_SIZE(gatt_delayed_store.peer_list); i++) {
+		/* Checking for the flags is cheaper than a memcmp for the
+		 * address, so we use that to signal that a given slot is
+		 * free.
+		 */
+		if (atomic_get(gatt_delayed_store.peer_list[i].flags) != 0) {
 			return false;
 		}
 	}
 
 	return true;
 }
 
-static void ccc_delayed_store(struct k_work *work)
+static void delayed_store(struct k_work *work)
 {
+	struct ds_peer *el;
 	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
-	struct gatt_ccc_store *ccc_store =
-		CONTAINER_OF(dwork, struct gatt_ccc_store, work);
+	struct gatt_delayed_store *store =
+		CONTAINER_OF(dwork, struct gatt_delayed_store, work);
 
-	for (size_t i = 0; i < CONFIG_BT_MAX_CONN; i++) {
-		struct bt_conn *conn = ccc_store->conn_list[i];
+	for (size_t i = 0; i < ARRAY_SIZE(gatt_delayed_store.peer_list); i++) {
+		el = &store->peer_list[i];
 
-		if (!conn) {
-			continue;
-		}
+		if (bt_addr_le_is_bonded(el->id, &el->peer)) {
+			if (IS_ENABLED(CONFIG_BT_SETTINGS_CCC_STORE_ON_WRITE) &&
+			    atomic_test_bit(el->flags, DELAYED_STORE_CCC)) {
+				atomic_clear_bit(el->flags, DELAYED_STORE_CCC);
+				bt_gatt_store_ccc(el->id, &el->peer);
+			}
 
-		if (bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
-			ccc_store->conn_list[i] = NULL;
-			bt_gatt_store_ccc(conn->id, &conn->le.dst);
-			bt_conn_unref(conn);
+			if (IS_ENABLED(CONFIG_BT_SETTINGS_CF_STORE_ON_WRITE) &&
+			    atomic_test_bit(el->flags, DELAYED_STORE_CF)) {
+				atomic_clear_bit(el->flags, DELAYED_STORE_CF);
+				bt_gatt_store_cf(el->id, &el->peer);
+			}
+
+			if (atomic_get(el->flags) == 0) {
+				gatt_delayed_store_free(el);
+			}
```

Comment:
Not sure if we need to do it with current implementation. During allocation we can override the `gatt_delayed_store` object if flags are set to zeros anyway

Reply:
let's do it just for correctness.